### PR TITLE
Added sim support

### PR DESCRIPTION
**Pull request by artemrad**
# Existing Sample Changes
## Description

Added simulator support to some tests

## External Dependencies
Dependent on the simulator

## Type of change

Please delete options that are not relevant. Add a 'X' to the one that is applicable. 

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Implement fixes for ONSAM Jiras

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [X] Command Line
- [ ] oneapi-cli
- [ ] Visual Studio
- [ ] Eclipse IDE
- [ ] VSCode
- [ ] When compiling the compliler flag "-Wall -Wformat-security -Werror=format-security" was used

### CODE REVIEW

**Comment by artemrad** [comment]
Tested: 
Linux: 
kernel_args_restrict: https://psg-sc-arc.sc.intel.com/arc/dashboard/reports/show_job/633435632
mem_channel: https://psg-sc-arc.sc.intel.com/arc/dashboard/reports/show_job/633395696
lsu_contrsol: https://psg-sc-arc.sc.intel.com/arc/dashboard/reports/show_job/633343364

Windows: 
kernel_args_restrict: https://psg-sc-arc.sc.intel.com/arc/dashboard/reports/show_job/634342459
mem_channel: https://psg-sc-arc.sc.intel.com/arc/dashboard/reports/show_job/634322172
lsu_contrsol: https://psg-sc-arc.sc.intel.com/arc/dashboard/reports/show_job/634225660

**Comment by artemrad** [comment]
@yuguen-intel  please review.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/kernel_args_restrict/src/CMakeLists.txt [hunk 1]
```diff
@@ -25,9 +25,10 @@ endif()
 #    For this reason, FPGA backend flags must be passed as link flags in CMake.
 set(EMULATOR_COMPILE_FLAGS "-fsycl -Wall ${WIN_FLAG} -fintelfpga -DFPGA_EMULATOR")
 set(EMULATOR_LINK_FLAGS "-fsycl -fintelfpga")
-set(SIMULATOR_COMPILE_FLAGS "-fsycl -Wall ${WIN_FLAG} -fintelfpga -DFPGA_SIMULATOR")
-set(SIMULATOR_LINK_FLAGS "-fsycl -fintelfpga -Xssimulation -Xsghdl -Xstarget=${FPGA_DEVICE} ${USER_HARDWARE_FLAGS}")
-set(HARDWARE_COMPILE_FLAGS "-fsycl -Wall ${WIN_FLAG} -fintelfpga")
+set(SIMULATOR_COMPILE_FLAGS "-Wall ${WIN_FLAG} -fsycl -fintelfpga -Xssimulation -DFPGA_SIMULATOR ${USER_SIMULATOR_FLAGS}")
+set(SIMULATOR_LINK_FLAGS "-fsycl -fintelfpga -Xssimulation -Xstarget=${FPGA_DEVICE} ${USER_SIMULATOR_FLAGS}")
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/kernel_args_restrict/src/CMakeLists.txt:29]: Please use the USER_HARDWARE_FLAGS for both the simulation and HW

File: DirectProgramming/DPC++FPGA/Tutorials/Features/kernel_args_restrict/src/CMakeLists.txt [hunk 2]
```diff
@@ -45,6 +46,20 @@ set_target_properties(${EMULATOR_TARGET} PROPERTIES COMPILE_FLAGS "${EMULATOR_CO
 set_target_properties(${EMULATOR_TARGET} PROPERTIES LINK_FLAGS "${EMULATOR_LINK_FLAGS}")
 add_custom_target(fpga_emu DEPENDS ${EMULATOR_TARGET})
 
+###############################################################################
+### FPGA Simulator
+###############################################################################
+# To compile in a single command:
+#    dpcpp -fintelfpga -Xssimulation -DFPGA_SIMULATOR kernel_args_restrict.cpp -o kernel_args_restrict.fpga_sim
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/kernel_args_restrict/src/CMakeLists.txt:53]:
```suggestion
#    icpx -fsycl -fintelfpga -Xssimulation -DFPGA_SIMULATOR kernel_args_restrict.cpp -o kernel_args_restrict.fpga_sim
```

File: DirectProgramming/DPC++FPGA/Tutorials/Features/kernel_args_restrict/src/CMakeLists.txt [hunk 3]
```diff
@@ -45,6 +46,20 @@ set_target_properties(${EMULATOR_TARGET} PROPERTIES COMPILE_FLAGS "${EMULATOR_CO
...
+# CMake executes:
+#    [compile] dpcpp -fintelfpga -Xssimulation -DFPGA_SIMULATOR -o kernel_args_restrict.cpp.o -c kernel_args_restrict.cpp
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/kernel_args_restrict/src/CMakeLists.txt:55]:
```suggestion
#    [compile] icpx -fsycl -fintelfpga -Xssimulation -DFPGA_SIMULATOR -o kernel_args_restrict.cpp.o -c kernel_args_restrict.cpp
```

File: DirectProgramming/DPC++FPGA/Tutorials/Features/kernel_args_restrict/src/CMakeLists.txt [hunk 4]
```diff
@@ -45,6 +46,20 @@ set_target_properties(${EMULATOR_TARGET} PROPERTIES COMPILE_FLAGS "${EMULATOR_CO
...
+#    [link]    dpcpp -fintelfpga -Xssimulation  kernel_args_restrict.cpp.o -o kernel_args_restrict.fpga_sim
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/kernel_args_restrict/src/CMakeLists.txt:56]:
```suggestion
#    [link]    icpx -fsycl -fintelfpga -Xssimulation  kernel_args_restrict.cpp.o -o kernel_args_restrict.fpga_sim
```

File: DirectProgramming/DPC++FPGA/Tutorials/Features/lsu_control/src/CMakeLists.txt [hunk 5]
```diff
@@ -47,6 +52,20 @@ set_target_properties(${EMULATOR_TARGET} PROPERTIES COMPILE_FLAGS "${EMULATOR_CO
 set_target_properties(${EMULATOR_TARGET} PROPERTIES LINK_FLAGS "${EMULATOR_LINK_FLAGS}")
 add_custom_target(fpga_emu DEPENDS ${EMULATOR_TARGET})
 
+###############################################################################
+### FPGA Simulator
+###############################################################################
+# To compile in a single command:
+#    dpcpp -fintelfpga -Xssimulation -DFPGA_SIMULATOR lsu_control.cpp -o lsu_control.fpga_sim
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/lsu_control/src/CMakeLists.txt:59]:
```suggestion
#    icpx -fsycl -fintelfpga -Xssimulation -DFPGA_SIMULATOR lsu_control.cpp -o lsu_control.fpga_sim
```

File: DirectProgramming/DPC++FPGA/Tutorials/Features/lsu_control/src/CMakeLists.txt [hunk 6]
```diff
@@ -47,6 +52,20 @@ set_target_properties(${EMULATOR_TARGET} PROPERTIES COMPILE_FLAGS "${EMULATOR_CO
...
+# CMake executes:
+#    [compile] dpcpp -fintelfpga -Xssimulation -DFPGA_SIMULATOR -o lsu_control.cpp.o -c lsu_control.cpp
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/lsu_control/src/CMakeLists.txt:61]:
```suggestion
#    [compile] icpx -fsycl -fintelfpga -Xssimulation -DFPGA_SIMULATOR -o lsu_control.cpp.o -c lsu_control.cpp
```

File: DirectProgramming/DPC++FPGA/Tutorials/Features/lsu_control/src/CMakeLists.txt [hunk 7]
```diff
@@ -47,6 +52,20 @@ set_target_properties(${EMULATOR_TARGET} PROPERTIES COMPILE_FLAGS "${EMULATOR_CO
...
+#    [link]    dpcpp -fintelfpga -Xssimulation  lsu_control.cpp.o -o lsu_control.fpga_sim
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/lsu_control/src/CMakeLists.txt:62]:
```suggestion
#    [link]    icpx -fsycl -fintelfpga -Xssimulation  lsu_control.cpp.o -o lsu_control.fpga_sim
```

File: DirectProgramming/DPC++FPGA/Tutorials/Features/lsu_control/src/CMakeLists.txt [hunk 8]
```diff
@@ -24,11 +25,15 @@ endif()
 #    For this reason, FPGA backend flags must be passed as link flags in CMake.
 set(EMULATOR_COMPILE_FLAGS "-Wall ${WIN_FLAG} -fsycl -fintelfpga -DFPGA_EMULATOR")
 set(EMULATOR_LINK_FLAGS "-fsycl -fintelfpga")
+set(SIMULATOR_COMPILE_FLAGS "-Wall ${WIN_FLAG} -fsycl -fintelfpga -Xssimulation -DFPGA_SIMULATOR ${USER_SIMULATOR_FLAGS}")
+# use cmake -D USER_SIMULATOR_FLAGS=<flags> to set extra flags for FPGA simulator compilation
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/lsu_control/src/CMakeLists.txt:29]: Please use USER_HARDWARE_FLAGS for both simulation and HW

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channel/src/CMakeLists.txt [hunk 9]
```diff
@@ -25,6 +26,9 @@ endif()
 #    For this reason, FPGA backend flags must be passed as link flags in CMake.
 set(EMULATOR_COMPILE_FLAGS "-Wall ${WIN_FLAG} -fsycl -fintelfpga -DFPGA_EMULATOR")
 set(EMULATOR_LINK_FLAGS "-fsycl -fintelfpga")
+set(SIMULATOR_COMPILE_FLAGS "-Wall ${WIN_FLAG} -fsycl -fintelfpga -Xssimulation -DFPGA_SIMULATOR ${USER_SIMULATOR_FLAGS}")
+set(SIMULATOR_LINK_FLAGS "-fsycl -fintelfpga -Xssimulation -Xstarget=${FPGA_DEVICE} ${USER_SIMULATOR_FLAGS}")
+# use cmake -D USER_SIMULATOR_FLAGS=<flags> to set extra flags for FPGA simulator compilation
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channel/src/CMakeLists.txt:31]: Please use USER_HARDWARE_FLAGS for both simulation and HW

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channel/src/CMakeLists.txt [hunk 10]
```diff
@@ -50,6 +54,20 @@ set_target_properties(${EMULATOR_TARGET} PROPERTIES COMPILE_FLAGS "${EMULATOR_CO
 set_target_properties(${EMULATOR_TARGET} PROPERTIES LINK_FLAGS "${EMULATOR_LINK_FLAGS}")
 add_custom_target(fpga_emu DEPENDS ${EMULATOR_TARGET})
 
+###############################################################################
+### FPGA Simulator
+###############################################################################
+# To compile in a single command:
+#    dpcpp -fintelfpga -Xssimulation -DFPGA_SIMULATOR mem_channel.cpp -o mem_channel.fpga_sim
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channel/src/CMakeLists.txt:61]:
```suggestion
#    icpx -fsycl -fintelfpga -Xssimulation -DFPGA_SIMULATOR mem_channel.cpp -o mem_channel.fpga_sim
```

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channel/src/CMakeLists.txt [hunk 11]
```diff
@@ -50,6 +54,20 @@ set_target_properties(${EMULATOR_TARGET} PROPERTIES COMPILE_FLAGS "${EMULATOR_CO
...
+# CMake executes:
+#    [compile] dpcpp -fintelfpga -Xssimulation -DFPGA_SIMULATOR -o mem_channel.cpp.o -c mem_channel.cpp
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channel/src/CMakeLists.txt:63]:
```suggestion
#    [compile] icpx -fsycl -fintelfpga -Xssimulation -DFPGA_SIMULATOR -o mem_channel.cpp.o -c mem_channel.cpp
```

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channel/src/CMakeLists.txt [hunk 12]
```diff
@@ -50,6 +54,20 @@ set_target_properties(${EMULATOR_TARGET} PROPERTIES COMPILE_FLAGS "${EMULATOR_CO
...
+#    [link]    dpcpp -fintelfpga -Xssimulation  mem_channel.cpp.o -o mem_channel.fpga_sim
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channel/src/CMakeLists.txt:64]:
```suggestion
#    [link]    icpx -fsycl -fintelfpga -Xssimulation  mem_channel.cpp.o -o mem_channel.fpga_sim
```

**Review by yuguen-intel — CHANGES_REQUESTED** [review]
A few small changes, you can just edit a reping me so that I merge this change

File: DirectProgramming/DPC++FPGA/Tutorials/Features/kernel_args_restrict/src/CMakeLists.txt [hunk 13]
```diff
@@ -25,8 +25,8 @@ endif()
 #    For this reason, FPGA backend flags must be passed as link flags in CMake.
 set(EMULATOR_COMPILE_FLAGS "-fsycl -Wall ${WIN_FLAG} -fintelfpga -DFPGA_EMULATOR")
 set(EMULATOR_LINK_FLAGS "-fsycl -fintelfpga")
-set(SIMULATOR_COMPILE_FLAGS "-Wall ${WIN_FLAG} -fsycl -fintelfpga -Xssimulation -DFPGA_SIMULATOR ${USER_SIMULATOR_FLAGS}")
-set(SIMULATOR_LINK_FLAGS "-fsycl -fintelfpga -Xssimulation -Xstarget=${FPGA_DEVICE} ${USER_SIMULATOR_FLAGS}")
+set(SIMULATOR_COMPILE_FLAGS "-Wall ${WIN_FLAG} -fsycl -fintelfpga -Xssimulation -DFPGA_SIMULATOR")
+set(SIMULATOR_LINK_FLAGS "-fsycl -fintelfpga -Xssimulation -Xstarget=${FPGA_DEVICE} ${USER_HARDWARE_FLAGS}")
 # use cmake -D USER_SIMULATOR_FLAGS=<flags> to set extra flags for FPGA simulator compilation
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/kernel_args_restrict/src/CMakeLists.txt:30]: You forgot to remove this comment in the three CMakeLists

**Review by yuguen-intel — CHANGES_REQUESTED** [review]

File: DirectProgramming/DPC++FPGA/Tutorials/Features/kernel_args_restrict/src/CMakeLists.txt [hunk 14]
```diff
@@ -25,9 +25,10 @@ endif()
 #    For this reason, FPGA backend flags must be passed as link flags in CMake.
 set(EMULATOR_COMPILE_FLAGS "-fsycl -Wall ${WIN_FLAG} -fintelfpga -DFPGA_EMULATOR")
 set(EMULATOR_LINK_FLAGS "-fsycl -fintelfpga")
-set(SIMULATOR_COMPILE_FLAGS "-fsycl -Wall ${WIN_FLAG} -fintelfpga -DFPGA_SIMULATOR")
-set(SIMULATOR_LINK_FLAGS "-fsycl -fintelfpga -Xssimulation -Xsghdl -Xstarget=${FPGA_DEVICE} ${USER_HARDWARE_FLAGS}")
-set(HARDWARE_COMPILE_FLAGS "-fsycl -Wall ${WIN_FLAG} -fintelfpga")
+set(SIMULATOR_COMPILE_FLAGS "-Wall ${WIN_FLAG} -fsycl -fintelfpga -Xssimulation -DFPGA_SIMULATOR")
+set(SIMULATOR_LINK_FLAGS "-fsycl -fintelfpga -Xssimulation -Xstarget=${FPGA_DEVICE} ${USER_HARDWARE_FLAGS}")
+# use cmake -D USER_SIMULATOR_FLAGS=<flags> to set extra flags for FPGA simulator compilation
```
Comment (artemrad) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/kernel_args_restrict/src/CMakeLists.txt:30]:
```suggestion
# use cmake -D USER_HARDWARE_FLAGS=<flags> to set extra flags for FPGA simulator compilation
```

File: DirectProgramming/DPC++FPGA/Tutorials/Features/lsu_control/src/CMakeLists.txt [hunk 15]
```diff
@@ -24,11 +25,15 @@ endif()
 #    For this reason, FPGA backend flags must be passed as link flags in CMake.
 set(EMULATOR_COMPILE_FLAGS "-Wall ${WIN_FLAG} -fsycl -fintelfpga -DFPGA_EMULATOR")
 set(EMULATOR_LINK_FLAGS "-fsycl -fintelfpga")
+set(SIMULATOR_COMPILE_FLAGS "-Wall ${WIN_FLAG} -fsycl -fintelfpga -Xssimulation -DFPGA_SIMULATOR")
+# use cmake -D USER_SIMULATOR_FLAGS=<flags> to set extra flags for FPGA simulator compilation
```
Comment (artemrad) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/lsu_control/src/CMakeLists.txt:29]:
```suggestion
# use cmake -D USER_HARDWARE_FLAGS=<flags> to set extra flags for FPGA simulator compilation
```

File: DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channel/src/CMakeLists.txt [hunk 16]
```diff
@@ -25,6 +26,9 @@ endif()
 #    For this reason, FPGA backend flags must be passed as link flags in CMake.
 set(EMULATOR_COMPILE_FLAGS "-Wall ${WIN_FLAG} -fsycl -fintelfpga -DFPGA_EMULATOR")
 set(EMULATOR_LINK_FLAGS "-fsycl -fintelfpga")
+set(SIMULATOR_COMPILE_FLAGS "-Wall ${WIN_FLAG} -fsycl -fintelfpga -Xssimulation -DFPGA_SIMULATOR")
+set(SIMULATOR_LINK_FLAGS "-fsycl -fintelfpga -Xssimulation -Xstarget=${FPGA_DEVICE} ${USER_HARDWARE_FLAGS}")
+# use cmake -D USER_SIMULATOR_FLAGS=<flags> to set extra flags for FPGA simulator compilation
```
Comment (artemrad) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channel/src/CMakeLists.txt:31]:
```suggestion
# use cmake -D USER_HARDWARE_FLAGS=<flags> to set extra flags for FPGA simulator compilation
```

**Review by yuguen-intel — APPROVED** [review]

**Pull request merged**